### ISSUE TITLE
02CIATIMERS

### ISSUE THREAD

**Issue by dmolinagarcia** [dmolinagarcia/74HCT6526#12]
Timer pulse on original CIA is aligned with PHI2 falling edge. On 74HCT6526 is aligned with posedge. 

![image](https://user-images.githubusercontent.com/30756488/216846446-2fbf3270-ab43-4e6b-bbf4-ce32707de9db.png)

Fixing this could alter all IRQ timings! Be careful.

**Comment by dmolinagarcia** [comment]
I need to run several test, with TA underflows, TB underflows, TA/TB underflows.

**Comment by dmolinagarcia** [comment]
HCT fires the timer pulse half a cycle late, but has the same value at PHI2 falling edge, so it should be read-identical via the PB pins. Still IRQ fires at the same time, so it may not be worth fixing this.

I need to move counting earlier (I fire the pulse as soon as the timer overflows, so theres no room to move the pulse earlier, without altering the whole pipeline).

I've been trying for a couple of days, and the pipeline is so delicate, any changes breaks it.

Moreover, if I count at PHI2 falling edge, I may not read a stable value so a latch would be needed, adding to more ICs (8bit latches) to the board.

All other timings seem perfect, at least from a CPU POV. This will remaing as is for now.

**Comment by dmolinagarcia** [comment]
Note to self. If I ever go this route, I need to replace the 244 with 373. They will latch the timer output on PHI2 high, so, when PHI falls, they retain the previous value from the timer, while they do their thing.

This me be extended to all Boards. And this may be well my biggest failure :/

**Comment by dmolinagarcia** [comment]
Issue renamed to reflect reality. This issue is behind the failure of test 02CIATIMERS. At least that's my current feeling. Emulator tests of 02TIMERS shows that, the very cycle the timer is reloaded, ICR bit is set. (This is NEWCIA behaviour) on OLDCIA, TIMERA does this, but TIMERB ICR is enabled one cycle later.

So, my switch between old mode and new mode is wrong. I need to move both Timers interrupts one cycle earlier (Danger! I need to re-do the pipeline 😭) And the OLD vs NEW change is to delay TIMERB by one cycle, not the interrupt being fired.

I need the full 02 test capture with the logic analyzer, with some changes.
PB7 and PB6 should output timer pulses. This should be captured
PHI2, should be captured
/CS for the CIA, captured
/IRQ as a bonus
On the 74HCT6526, I can add ICR0 and ICR1 too.

I can infer the registers from the code so the are not needed. Capture single subtest at a time. We're looking at 20*200 cycles max per test, so it should be easy to compare.

What am I looking for? Exactness of the pulse (ok ok, I know It is delayed by half a cycle)

**Comment by dmolinagarcia** [comment]
OLD vs OLD (bottom result)

![image](https://user-images.githubusercontent.com/30756488/230973594-1e717929-f962-42cd-8e64-602029d5f78f.png)

![image](https://user-images.githubusercontent.com/30756488/230973610-db5ddcee-356b-441f-b3e5-8fadaf22941e.png)

NEW vs NEW (again, bottom result)

![image](https://user-images.githubusercontent.com/30756488/230973663-40a17bc7-e49d-4389-8fa2-9ee2087780a0.png)

![image](https://user-images.githubusercontent.com/30756488/230973674-99bd8b11-fdb3-4a9d-8845-670b4c406354.png)

**Comment by dmolinagarcia** [comment]
TIMER A interrupt shows no difference between new or old. It's timerB the one that has a difference. TIMERB interreupt seems to fire a cycle earlier on NEW cias.

My CIA behaves like an old one. Keep in mind, this is always relative to the ICR register. I am setting the ICR bit as soon as the overflow happens. there's  no room to fire earlier.

**Comment by dmolinagarcia** [comment]
Bite me! I'm getting it al wrong!

on first test, when reading ICR, I'm reading a 0 when there should be a 1, but not because I'm setting it half a cycle late... The 1 I should be reading, has fired long before, right at the end of cycle B.

I am clearing the ICR when I shouldn't. That's what I need to tackle. Issue now feels to exists only at ICR level.

**Comment by dmolinagarcia** [comment]
After analyzing the test line by line, my failure comes from clearing interrupts too late. So I miss already set interrupt, by creating the clear before it even triggers.

Moving the clear for TA one cycle earlier, I am getting TA with no IRQ to end with OK!

This is stage one. To change from NEW to OLD, Clear of TB needs to move from where it is now (OLD) to one cycle earlier (NEW)

Still some differences in the IRQ enabled tests. Full capture of both old and new (mos and hct9 is done. time to analyze it!

**Comment by dmolinagarcia** [comment]
STAGE 1

HCT-NEW
![image](https://user-images.githubusercontent.com/30756488/235541364-37bbd942-a1ab-4d0a-b796-dcd44640ba2a.png)

HCT-OLD
![image](https://user-images.githubusercontent.com/30756488/235541582-d3e2cc5d-176b-4e49-9e20-b9cf572e7a32.png)

MOS-NEW
![image](https://user-images.githubusercontent.com/30756488/235541778-263452e0-c979-460a-af56-616a747f57d3.png)

MOS-OLD
![image](https://user-images.githubusercontent.com/30756488/235541994-e9b206ed-0302-4198-99fe-0215d1da7839.png)

**Comment by dmolinagarcia** [comment]
Some changes to ICR have been applied.
STAGE0. Cleanup of the dangling flipflop. Glued to the board and with bodge wires replaced. Everything still works the same, but looks much nicer.
STAGE1. ICR0 reset moved one cycle earlier. Fix fixes test 5
STAGE2. Same for ICR0, however, OLD still resets after two, NEW after one
BONUS. This seems to be behind timerB bug.
STAGE3. ICR7 moved one cycle earlier for NEW mode. Works fine!

This is enough to pass 02NEW. 

02OLD still shows some issues, which I expect to fix in the next two stages
STAGE4. ICR CLEAR. Reset both ICR7 DFF at the same time. Right now, second one is delayed
STAGE5. IRQ. I have the feeling IRQ does not clear at the same time as ICR7, so I may need a new 74HC74 for this.

None of the above is applied to the schematic.. When S4 and S5 are done and validated, everything will be updated.

**Comment by dmolinagarcia** [comment]
Next steps.

Consolidate :
- Stage 1. Fixes TA clear. Same for both modes
- Stage 2. Switch for TB clear. /ICRCLR1 for NEW /ICRCLR2 for OLD
- Stage 3. ICR7 switch. One cycle earlier for NEW

This is confirmed as good.

Then...

Stage 4 alters the clear of ICR on OLD mode, clearing it one cycle earlier. This alters the IRQ in OLD too. NEW shouldn't be affected.
Stage 5 adds 2 new FF for IRQ. The set is same as ICR7, but clear is different. 

Adding 4 and 5 I'm getting weird results that I do not understand, so let's backtrack a bit.

- change the new FF so the clear is same as IC7. This should be the same as undoing stage 5.
- undo stage 4. Let's see if I get back to known conditions.

Once here, take capture of the test. Both new and old, HCT and MOS.

**Comment by dmolinagarcia** [comment]
Rollback to STAGE3. 02NEW success. But IRQ is produced in a separate set of FF than ICR7.

02OLD fails in two scenearios.
A read right before the interrupt for TB clears ICR1 but not ICR7, so a 0x80 is read. I tried moving ICR7 clear forward, but it remains set always and needs a second read to clear ICR7. I think triggering the ICR7 set with TBPULSE instead of ICR1 can do the trick. Try this later today.

Also, the next iteration, no interrupt should fire. This can probably by fixed ltering the clear for IRQ. 

Now that I have separated ICR7 and IRQ, this two issues are independent.

But, as everything checks out in NEW mode, it may be a good idea to make a commit (v0.2.2 I think) with NEW mode working. I really need to consolidate the schematic, as B3 is becoming more and more hard to follow.

**Comment by dmolinagarcia** [comment]
Hold your horses!

I've managed to succesfully apply a STAGE4 that fixes the first remaining issue:

A read right before the interrupt for TB clears ICR1 but not ICR7, so a 0x80 is read. I tried moving ICR7 clear forward, but it remains set always and needs a second read to clear ICR7. I think triggering the ICR7 set with TBPULSE instead of ICR1 can do the trick. Try this later today.

First of all. All interrupt sources must be cleaned by 

<span style="text-decoration:overline">SIGNAL_ACTIVE_LOW</span>
